openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/WorkflowInstanceStateRepository.java

```diff
@@ -176,6 +177,51 @@ public void updateStageWithTask(UUID taskId, UUID workflowInstanceStateId) {
     getTimeSeriesDao().update(JsonUtils.pojoToJson(workflowInstanceState), workflowInstanceStateId);
   }
 
+  /**
+   * Marks all states of a workflow instance as FAILED with the given reason.
+   * Preserves audit trail instead of deleting the states.
+   */
+  public void markInstanceStatesAsFailed(UUID workflowInstanceId, String reason) {
+    try {
+      // Get workflow definition to query states
+      WorkflowInstanceRepository workflowInstanceRepository =
+          (WorkflowInstanceRepository)
+              Entity.getEntityTimeSeriesRepository(Entity.WORKFLOW_INSTANCE);
+      WorkflowInstance instance =
+          JsonUtils.readValue(
+              workflowInstanceRepository.getTimeSeriesDao().getById(workflowInstanceId),
+              WorkflowInstance.class);
+
+      WorkflowDefinitionRepository workflowDefinitionRepository =
+          (WorkflowDefinitionRepository) Entity.getEntityRepository(Entity.WORKFLOW_DEFINITION);
+      WorkflowDefinition workflowDefinition =
+          workflowDefinitionRepository.get(
+              null, instance.getWorkflowDefinitionId(), EntityUtil.Fields.EMPTY_FIELDS);
+
+      // Query all states for this workflow instance
+      long endTs = System.currentTimeMillis();
+      long startTs = endTs - (7L * 24 * 60 * 60 * 1000);
+
+      ResultList<WorkflowInstanceState> instanceStates =
+          listWorkflowInstanceStateForInstance(
+              workflowDefinition.getName(), workflowInstanceId, null, startTs, endTs, 1000, false);
```

Comment:
these few lines about getting the state are used a couple of places. Would be good to have a common method so we don't have discrepancies by listing with different times in different places